### ISSUE TITLE
commandButton using onclick and action within ui:fragment

### ISSUE THREAD

**Issue by mydeadlyvenoms** [TheCoder4eu/BootsFaces-OSP#399]
After upgrading to BootsFaces 0.8.6 (from 0.8.1) my SPA seems to be broken.

**index.xhtml**
Within the index.xhtml I have a top-level navigation, something like:
```xml
<b:navbarLinks>
  <b:navLink value="Home" icon="home" iconAlign="left" onclick="ajax:contentBacking.setContent('home')" update="content" />
...
```
And here we have the dynamic content loaded using ajax:
```xml
<h:panelGroup id="content">
  <ui:include src="/WEB-INF/view/#{contentBacking.content}.xhtml" />
</h:panelGroup>
```
This does still work as expected.

**home.xhtml**
Within the ui:fragment loaded using ajax the navigation does only work after a second click, which is a new behavior since BootsFaces 0.8.6 is being used. This means I have to click the commandButton twice to trigger the onclick event.

```xml
<ui:fragment 
    xmlns="http://www.w3.org/1999/xhtml"
    xmlns:b="http://bootsfaces.net/ui"
    xmlns:f="http://xmlns.jcp.org/jsf/core"
    xmlns:h="http://xmlns.jcp.org/jsf/html"
    xmlns:ui="http://xmlns.jcp.org/jsf/facelets">

...

<h:form>
<b:commandButton value="BLUB" look="success" size="lg" iconAwesome="play" 
onclick="ajax:contentBacking.setContent('new-content-here')" update="content" ajax="true" />
</h:form>

...

</ui:fragment>
```

**Comment by pavankd** [comment]
Hi mydeadlyvenoms,
                             I too facing the same problem.Please help me

**Comment by zhedar** [comment]
Sadly I couldn't reproduce your exact problem. The Ajax request happened always on the first click, no matter if it is included or not.
Would it be possible for you to provide a full running minimal example?

**Comment by pavankd** [comment]
@zhedar  could you please provide example code

**Comment by zhedar** [comment]
@pavankd There is a [simple quickstart project](https://github.com/TheCoder4eu/BootsFaces-examples) you could use as a start to embed your own failing code, but it I've just seen, that it only has a gradle build file and lacks a maven pom. You could also always fork [our Showcase](https://github.com/TheCoder4eu/BootsFacesWeb) and include it there, then link to it. So that's about starting points for easily runnable example projects. Or didn't I understand you correctly?

**Comment by mydeadlyvenoms** [comment]
Hi there 
first of all
`<h:outputScript library="bsf" name="js/button.js" target="head"/>`
does not work... neither within the index.xhtml nor within the ui:fragment.

@pavankd do you have an on-the-fly minimal example? Otherwise I'll create one.

**Comment by pavankd** [comment]
@mydeadlyvenoms  no dont have any on-the-fly minimal example

**Comment by mydeadlyvenoms** [comment]
A little late.. but here you go:

[https://github.com/mydeadlyvenoms/BootsFacesIssues](https://github.com/mydeadlyvenoms/BootsFacesIssues)

The crazy thing is, that the minimal example behaves a little different, but the main problem still exists.

Please let me know if you need more information.

**Comment by mydeadlyvenoms** [comment]
@zhedar were you able to reproduce the issue? 
I just found out that in the minimal example the first button triggers directly while the second (form) doesn't. Within my original project this is not the case, every button inside a ui:fragment needs to be clicked twice. So it must have something to do with components i deleted for the minimal example.

**Comment by m1rc0** [comment]
Hi there,

I've probably the same issue. I've a navbar from which I navigate into different views. According to the ajax navigation example everything works fine. But in some cases I've a b:listLinks with multiple b:navLink in a view and then I've to click the first link twice to make it work.

I'm also on Payara. 

Thank you very much

**Comment by zhedar** [comment]
@mydeadlyvenoms I was able to reproduce it. The second click seems to be needed, when you navigated there via ajax. If you change to about and back to home, you will have to click twice as well.

**Comment by zhedar** [comment]
I should learn to read. 

>After upgrading to BootsFaces 0.8.6 (from 0.8.1)

Now that I know, that we actually broke tha, we can track down the commit, that did it. Then we might know, how we to solve this.

Just tested this with 0.8.1 and it works (partially, as there are errors in home-step-1.

**Comment by mydeadlyvenoms** [comment]
@zhedar you nailed it, thats exactly how it behaves when I test it. Nasty bug...

**Comment by m1rc0** [comment]
@mydeadlyvenoms Thank you. And @zhedar do you have an idea how to fix it?

Thank you very much

**Comment by mydeadlyvenoms** [comment]
@zhedar do you have any updates on this? 
Sadly it has not been tagged for 0.9. Maybe we need to remove the <verifying issue> label since it can be reproduced now.

**Comment by stephanrauh** [comment]
@zhedar Just tell me if you want me to investigate the bug tomorrow. I can't promise I'll solve it in time, but it looks important to me. What's your opinion?

**Comment by zhedar** [comment]
@stephanrauh I'm probably not at my desk for the next 2 days, so probably can't fix that. I can only say that it appeared first in 572e766, maybe that helps you. But that's about it, if it comes to things I know about this one ;(

**Comment by stephanrauh** [comment]
I tried the application on IBM Websphere Liberty. It worked fine. Two consecutive clicks lead me from HOME-STEP-0 over HOME-STEP-1 to HOME-STEP-2.

Running the same program on Wildfly show a completely different behavior. Starting from HOME-STEP-0: First click: nothing. Second click: HOME-STEP-1. Third click: HOME-STEP-0 again. The AJAX responses seem to be OK (no error message or something), they just have the wrong content.

Is that the error you've been describing?

**Comment by mydeadlyvenoms** [comment]
Hi there

Uff, interesting behavior...
This is exactly the Problem I wanted to show using the provided minimal example.
From time to time also some clicks on the main navigation only work on second click (after triggering some ajax based navigations within the ui:fragment), but until today I was not able to identify a pattern.

**Comment by stephanrauh** [comment]
There seem to be at least two bugs:

- The view scope of the `ContentBacking` bean seems to be corrupt. The third click I described above creates a new instance of the bean, which in turn explains why the application starts with the first page again.
- After setting the scope to session scope and populating the `value` attributes of the input field, I can reproduce the original error you describe. The first click causes an AJAX request, but the AJAX request doesn't change the page. Only the second AJAX request manages this. The difference between the requests is that the first request doesn't send the view state with the header.

**Comment by stephanrauh** [comment]
@mydeadlyvenoms @zhedar When I replaced the CDI annotations by the older JSF annotations (`@ManagedBean` and `javax.faces.bean.ViewScoped`), the program runs fine. May I ask you to confirm this?

**Comment by mydeadlyvenoms** [comment]
If I understood you right, you just replaced the annotations inside the ContentBacking?

```
import javax.faces.bean.ManagedBean;
import javax.faces.bean.ViewScoped;
```

and

```
@ManagedBean
@ViewScoped
public class ContentBacking {
...
```

does exactly the same (bugs are still present).

Please let me know if I can further assist you.

**Comment by stephanrauh** [comment]
You're right, I've changed two things at once. Try to wrap the `<ui:include />` in an `<h:form />`. That does the trick for me, and according to http://stackoverflow.com/questions/10094615/a4jcommandbutton-action-is-only-invoked-on-second-click, it may be a JSF bug.

```xml
       <h:panelGroup id="content" style="padding-top:50px">
          <h:form>
              <ui:include src="/WEB-INF/view/#{contentBacking.content}.xhtml" />
          </h:form>
        </h:panelGroup>
```
I just wonder what our poisoned commit did to make the bug worse?

**Comment by stephanrauh** [comment]
@zhedar @mydeadlyvenoms Currently, I fail to reproduce the regression. The bug happens even if I switch to BF 0.8.1 and 0.8.0. Maybe there's a subtly difference between Websphere and Glassfish?

**Comment by zhedar** [comment]
I kind of thought about the stacked form bug, but somehow didn't find proof, that this one could be affected by it.
@stephanrauh That's strange. So checking out a commit before the mentioned one still triggers the bug? Because I'm not sure I've actually checked those releases (although I think I checked atleast 0.8.1)
Also I'm currently using Wildfly 10 and therefore Mojarra 2.2.13.SP1. I will try to reproduce it again later today.

**Comment by stephanrauh** [comment]
@zhedar THX! I tried the releases from Maven Central. I also tried to checkout older source code version of BootsFaces, but that proved to be difficult because you have to deinstall the JSFDSL plugin first. So I wasn't sure if something went wrong during compilation. But I believe the Maven-Central-version-test was reliable enough.

**Comment by mydeadlyvenoms** [comment]
@stephanrauh thanks for the fast reply. Also `<h:form> `doesn't change anything... Have you tried your solution on GlassFish / Payara? I still have to click everything twice.

**Comment by zhedar** [comment]
@stephanrauh You have to deinstall the plugin? I normally check the commit out, then only trigger a `mvn clean install`. If eclipse mocks you, you can also temporarily disable the XText nature of the project under the Configure option when right clicking on the project.

**Comment by stephanrauh** [comment]
@mydeadlyvenoms Last time I checked Glassfish still was buggy as hell. OK, that's two or three years ago, so I've downloaded it and try to start it. No luck. Eclipse says that Payara doesn't support CDI. It seems neither Glassfish nor Payara want to be friends of mine :(.

**Comment by zhedar** [comment]
I've tried WildFly 10 and Payara 4.1 so far and can confirm @stephanrauh's fix is working there. I removed the form from the include and put it in the template instead:

     <h:panelGroup id="content">
          <h:form>
             <ui:include src="/WEB-INF/view/#{contentBacking.content}.xhtml" />
          </h:form>
     </h:panelGroup>

Everything seem to work as it should. However I'm still not sure, how the hell we even broke that. Also Payara was quite a hassle to get running, but I got it to play with Eclipse.

**Comment by zhedar** [comment]
Oh well, that only applies to the bug we had with the different sites. The original twice click-thingy is still present.

Steps to reproduce:
* switch to `About` 
* switch back to `Home`
* click `Launch Instance` button
* rage, because you need to click again

**Comment by zhedar** [comment]
Good news, everyone.

I could fix the twice click problem by putting everything in a single form and updating this form via the `@form` search expression. That still doesn't explain, how we broke it, but it seems to be a JSF related bug. I guess using search expressions (e.g. those from PrimeFaces, but ours should do as well) may be a solution here.

**Comment by mydeadlyvenoms** [comment]
Hi everyone, thank you very much for your help.

@zhedar so the solution for now is to put everything (including the navBar) into one sigle `h:form` and then always update the view using `@form`? Doesn't this approach introduce lots of overhead?

In the long run I don't like the idea of using legacy annotations and imports. Currently it isn't clear to me if this is a BootsFaces oder a general JSF issue? Absolutely no offense, I'm just wondering. :-)

**Comment by zhedar** [comment]
I'm sure there are other ways, but that should work in any case. The overhead shouldn't be that high, if you're not using a humongous form. 

Also you don't have to use the old imports. I haven't changed the annotations(I'm guessing you used the ones from CDI). I guess they don't have to do anything with this particular problem.

**Comment by stephanrauh** [comment]
As far as I can see, you don't have to update the entire form. The only thing that's important is that the `<h:form>` tag is not in the include file but in the root file. I suspect JSF simply doesn't recognize the form as a form when it sends the AJAX request if you do otherwise. In any case, you can combine `process="@form" update="@form:content"` to reduce the size of the response.

**Comment by stephanrauh** [comment]
Damiano, I'd like to close the issue soon. Does you SPA work now?

Todos left:
- Document SPAs in the showcase
- update the BeyondJava.net article on BootsFaces SPAs

**Comment by m1rc0** [comment]
@zhedar and @stephanrauh Thank you very much for your time and effort. 

I created another example here https://github.com/m1rc0/BootsFacesExample and for the moment the workaround with putting everything inside a single h:form seems to work. But I agree with @mydeadlyvenoms that that is not a very elegant solution.

**Comment by mydeadlyvenoms** [comment]
Hi everyone

I can confirm that everything works now, putting all the stuff into one "parent" form solved all the issues. @m1rc0 the cool thing is that you don't have to mess with forms within the imports now. Since I can update the content using `update="@form:content"` during the navigation I am okay with that.

So my include wrapper looks like this now:

```
<h:panelGroup id="content">
  <ui:include src="/WEB-INF/view/#{contentBacking.content}.xhtml" />
</h:panelGroup>
```

**Comment by mydeadlyvenoms** [comment]
What I forgot to mention:

Since all the imports do not contain forms anymore, I can now remove OmniFaces `fixviewstate.js`.

**Comment by stephanrauh** [comment]
@mydeadlyvenoms @m1rc0 I'm very glad this workaround does the trick for you! Of course, I'd prefer to know how to fix the bug, but that may be difficult.

Does anybody object closing the issue?

**Comment by mydeadlyvenoms** [comment]
I think you can close the issue.
